### PR TITLE
ROOM 4: graph neighbourhood added

### PR DESCRIPTION
**Pull request by leriomaggio**
This PR will add to the repo a Python module to determine the list of arcs in a graph whose weight is greater or equal to their corresponding neighbour coordinates.

The module includes a `neighbourhood` function that determines the neighbourhood of each node which is independent, and importable from other third-party code. 

The module _also_ contains a **main** section to run and test the program that is working with some input file representing the connections between nodes in a graph.

**Note**:
I'd like to receive comments and suggestions about the organisation of the code, and whether you think my code could be improved in *any* way. 
This is my **first** contribution to the project, and I am not very experienced in Python. 
I've read guidelines on how to contribute to the repo, but I'd love to receive any comments you might have.

### CODE REVIEW

File: snippets/graph_neighbourhood.py [hunk 1]
```diff
@@ -0,0 +1,68 @@
+""""""
```
Comment (krassowska) [inline snippets/graph_neighbourhood.py:1]: Please add description in the docstring of what your code is doing.

Comment (ykhorrami) [inline snippets/graph_neighbourhood.py:1]: Need a good name in the start of the code

File: snippets/graph_neighbourhood.py [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+        (0, 1),
+        (-1, -1),
+        (-1, 1),
+        (1, -1),
+        (1, 1),
+    ]
+    return filter(
+        lambda l: and_(*starmap(le, zip((0, 0), l)))
+        and and_(*starmap(lt, zip(l, borders))),
+        map(lambda n: tuple(map(sum, zip(coord, n))), (n, s, w, e, nw, ne, sw, se)),
```
Comment (jamesrharwood) [inline snippets/graph_neighbourhood.py:27]: Lambda functions can be difficult to understand for a reader. I would prefer it if you created small functions whose names describe what the function is doing.

File: snippets/graph_neighbourhood.py [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+from itertools import starmap
+from operator import le, lt, and_
+from pathlib import Path
+from os import path
+from sys import argv, stdout
+from argparse import ArgumentParser
+
+from data import load_graph
+
+
```
Comment (ykhorrami) [inline snippets/graph_neighbourhood.py:12]: There is more space here.

File: snippets/graph_neighbourhood.py [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+        (-1, -1),
+        (-1, 1),
+        (1, -1),
+        (1, 1),
+    ]
+    return filter(
+        lambda l: and_(*starmap(le, zip((0, 0), l)))
+        and and_(*starmap(lt, zip(l, borders))),
+        map(lambda n: tuple(map(sum, zip(coord, n))), (n, s, w, e, nw, ne, sw, se)),
+    )
```
Comment (steven-wooding) [inline snippets/graph_neighbourhood.py:28]: I find it hard to follow the code without any explanatory comments - but that is (largely?) because I don't program in Python.

File: snippets/graph_neighbourhood.py (hunk 2 above)
Comment (jamesrharwood) [inline snippets/graph_neighbourhood.py:27]: I would find this clearer if you wrapped this filter call in its own function whose name describes what it is doing. Then return the result of that function call.

File: snippets/graph_neighbourhood.py [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+from operator import le, lt, and_
+from pathlib import Path
+from os import path
+from sys import argv, stdout
+from argparse import ArgumentParser
+
+from data import load_graph
+
+
+def neighbourhood(coord: tuple[int, int], borders) -> None:
```
Comment (jamesrharwood) [inline snippets/graph_neighbourhood.py:13]: Please could you define what borders should be?

Comment (jamesrharwood) [inline snippets/graph_neighbourhood.py:13]: Looks like you are expecting the function to return None, but the return statement at the end contradicts this

File: snippets/graph_neighbourhood.py [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+    n, s, w, e, nw, ne, sw, se = [
+        (-1, 0),
+        (1, 0),
+        (0, -1),
+        (0, 1),
+        (-1, -1),
+        (-1, 1),
+        (1, -1),
+        (1, 1),
+    ]
```
Comment (ykhorrami) [inline snippets/graph_neighbourhood.py:23]: It's better to define a "list" to show the numbers or we can use an algorithm in another def to create these numbers like (-1,0), (1,0), ....

File: snippets/graph_neighbourhood.py [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+    return filter(
+        lambda l: and_(*starmap(le, zip((0, 0), l)))
+        and and_(*starmap(lt, zip(l, borders))),
+        map(lambda n: tuple(map(sum, zip(coord, n))), (n, s, w, e, nw, ne, sw, se)),
+    )
+
+
+if __name__ == "__main__":
+    parser = ArgumentParser(
+        description="Determine the list of arcs in the input Graph whose weight is greater or equal to their neighbour coordinates."
```
Comment (steven-wooding) [inline snippets/graph_neighbourhood.py:33]: Could you clarify whether an 'arc' is another term for 'edge' or has some special technical meaning - ie is it a special type of 'edge'.

File: snippets/graph_neighbourhood.py [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    )
```
Comment (ykhorrami) [inline snippets/graph_neighbourhood.py:34]: I prefer to define a variable for "Determine the list of arcs in the input Graph whose weight is greater or equal to their neighbour coordinates" out of the if condition.

File: snippets/graph_neighbourhood.py [hunk 9]
```diff
@@ -0,0 +1,68 @@
...
+    if (
+        cli_args.logfile_output_path is not None
+        and cli_args.logfile_output_path.exists()
+    ):
+        logfile = open(cli_args.logfile_output_path, "w")
+    else:
+        logfile = stdout
+
+    for edge in nn:
+        print(f"{edge}", file=logfile)
```
Comment (steven-wooding) [inline snippets/graph_neighbourhood.py:68]: Is it possible to pass a file path to 'logfile' or is only a filename acceptable - it might be worth renaming as logfilepath if a file path is acceptable.

File: snippets/graph_neighbourhood.py [hunk 10]
```diff
@@ -0,0 +1,68 @@
...
+        (1, 1),
+    ]
+    return filter(
+        lambda l: and_(*starmap(le, zip((0, 0), l)))
+        and and_(*starmap(lt, zip(l, borders))),
+        map(lambda n: tuple(map(sum, zip(coord, n))), (n, s, w, e, nw, ne, sw, se)),
+    )
+
+
+if __name__ == "__main__":
```
Comment (jamesrharwood) [inline snippets/graph_neighbourhood.py:31]: I would consider splitting this into smaller functions. At the moment it's not clear to me what this is going to do

File: snippets/graph_neighbourhood.py [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    cli_args = parser.parse_args()
+    adj_matrix = load_graph(cli_args.data_filepath, sparse=False, fillna_with=0)
+    N = len(adj_matrix)
+
+    nn = list()
+    for u, _ in enumerate(adj_matrix):
+        for v, w in enumerate(adj_matrix[u]):
+            for (i, j) in neighbourhood((u, v), borders=(N, N)):
+                if w >= adj_matrix[i][j]:
+                    nn.append((u, v))
```
Comment (jamesrharwood) [inline snippets/graph_neighbourhood.py:56]: Could you use descriptive variable names?

**Review by steven-wooding — COMMENTED** [review]
It's hard to review Python code when you don't program in Python.

File: snippets/graph_neighbourhood.py [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+    print(f"Found {len(nn)} edges in the input Graph satisfying the condition.")
+
+    if (
+        cli_args.logfile_output_path is not None
+        and cli_args.logfile_output_path.exists()
```
Comment (jamesrharwood) [inline snippets/graph_neighbourhood.py:61]: I would consider wrapping this into a small function that validates whether a valid file path has been provided. Then you could use the same function to validate the input file path too.

File: snippets/graph_neighbourhood.py [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+    )
+    parser.add_argument(
+        "-o", "--output", type=Path, dest="logfile_output_path", default=None
+    )
+
+    cli_args = parser.parse_args()
+    adj_matrix = load_graph(cli_args.data_filepath, sparse=False, fillna_with=0)
+    N = len(adj_matrix)
+
+    nn = list()
```
Comment (DLR-MarkusSchreiber) [inline snippets/graph_neighbourhood.py:51]: This should be put in a callable function as you described it in the pull request description.

Comment (DLR-MarkusSchreiber) [inline snippets/graph_neighbourhood.py:51]: Try to use readable variable names.

File: snippets/graph_neighbourhood.py [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+            for (i, j) in neighbourhood((u, v), borders=(N, N)):
+                if w >= adj_matrix[i][j]:
+                    nn.append((u, v))
+    print(f"Found {len(nn)} edges in the input Graph satisfying the condition.")
+
+    if (
+        cli_args.logfile_output_path is not None
+        and cli_args.logfile_output_path.exists()
+    ):
+        logfile = open(cli_args.logfile_output_path, "w")
```
Comment (jamesrharwood) [inline snippets/graph_neighbourhood.py:63]: Consider:
- whether you want to 'append' instead of 'write'
- you don't close this file. Consider whether you want to refactor so that the write statement is called within a `with` statement

File: snippets/graph_neighbourhood.py [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+    parser.add_argument(
+        "-o", "--output", type=Path, dest="logfile_output_path", default=None
+    )
+
+    cli_args = parser.parse_args()
+    adj_matrix = load_graph(cli_args.data_filepath, sparse=False, fillna_with=0)
+    N = len(adj_matrix)
+
+    nn = list()
+    for u, _ in enumerate(adj_matrix):
```
Comment (niketagrawal) [inline snippets/graph_neighbourhood.py:52]: Please consider encapsulating the logic in the for loop in a function with a relevant name. This will help improve the readability of the code.

File: snippets/graph_neighbourhood.py [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+from pathlib import Path
+from os import path
+from sys import argv, stdout
+from argparse import ArgumentParser
+
+from data import load_graph
+
+
+def neighbourhood(coord: tuple[int, int], borders) -> None:
+    n, s, w, e, nw, ne, sw, se = [
```
Comment (niketagrawal) [inline snippets/graph_neighbourhood.py:14]: Please rename the variables to give them a more meaningful context. For e.g., 'n' could be renamed to north, nw could be renamed to northwest assuming that these variables represent the directions.

File: snippets/graph_neighbourhood.py (hunk 5 above)
Comment (niketagrawal) [inline snippets/graph_neighbourhood.py:13]: Including a docstring for the function renders readability to the code. You can consider using the [Numpy-doc standard](https://numpydoc.readthedocs.io/en/latest/format.html#docstring-standard) for writing a docstring. You can reuse the function description in the PR to put in the docstring.

Good to see that the type is annotated for the first function argument; you could do the same for the second argument 'borders' as well.

File: snippets/graph_neighbourhood.py (hunk 6 above)
Comment (leriomaggio) [reply snippets/graph_neighbourhood.py:23]: Well, you're totally right and you could actually look at `itertools.product` in standard Python library to generate those coordinates.

Comment (leriomaggio) [reply snippets/graph_neighbourhood.py:23]: What I meant was something like
```
from itertools import product 
coordinates = product((-1, 0, 1), 2)
```
